client/allocrunner/consul_hook.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package allocrunner
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/consul"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/client/widmgr"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// consulTokenFilePrefix is the begging of the name of the file holding the
+	// Consul SI token inside the task's secret directory. Full name of the file is
+	// always consulTokenFilePrefix_identityName
+	consulTokenFilePrefix = "nomad_consul_"
+
+	// consulTokenFilePerms is the level of file permissions granted on the file in
+	// the secrets directory for the task
+	consulTokenFilePerms = 0440
+)
+
+type consulHook struct {
+	alloc         *structs.Allocation
+	allocdir      *allocdir.AllocDir
+	widmgr        widmgr.IdentityManager
+	client        consul.Client
+	hookResources *cstructs.AllocHookResources
+	authMethod    string
+
+	logger log.Logger
+}
+
+func newConsulHook(logger log.Logger, alloc *structs.Allocation,
+	allocdir *allocdir.AllocDir,
+	widmgr widmgr.IdentityManager,
+	client consul.Client,
+	hookResources *cstructs.AllocHookResources,
+	authMethod string,
+) *consulHook {
+	h := &consulHook{
+		alloc:         alloc,
+		allocdir:      allocdir,
+		widmgr:        widmgr,
+		client:        client,
+		hookResources: hookResources,
+		authMethod:    authMethod,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*consulHook) Name() string {
+	return "consul"
+}
+
+func (h *consulHook) Prerun() error {
+	job := h.alloc.Job
+
+	if job == nil {
+		// something crazy happened
+		err := fmt.Errorf("alloc %v does not have a job", h.alloc.Name)
+		h.logger.Error(err.Error())
+		return err
+	}
+
+	mErr := multierror.Error{}
+	tokens := map[string]string{}
+
+	for _, tg := range job.TaskGroups {
+		for _, task := range tg.Tasks {
+			req, err := h.prepareConsulClientReq(task)
```

Comment:
This method handles services defined at the task level, but we'll need to handle the services defined at the group level too.

Reply:
Right! Addressed in 06536f9474ed02ced4a1fba0ff8ef89630af2122